doc/damnit.man
.TH DAMNIT _MANSEC_
.SH NAME
damnit \- DAtabase MaNagement InTerface

.SH SYNOPSIS
.B damnit 
[\-h] 
[\-d\ <file>\ <value>]
[\-a\ <file>\ <value>] 
[\-f\ <string>]
[\-l\ <file>]

.SH DESCRIPTION
.B damnit
is the database management tool for 
.B gnarwl(_MANSEC_)
\&. It allows
systemadministrators to list and/or manipulate 
.B gnarwl's
database files. It is 
not intented to be employed by the average user, who should use LDAP as
the only interface to
.B gnarwl(_MANSEC_)
.

.SH OPTIONS
.IP -h
Print usage information

.IP "-f <string>"
Select output format for database listing (only meaningful with -l).
<string> is the template for what the output should look like. The following
macros are recognized: %entry, %time, %tstamp, \\en and \\et. Translating to
the entry itself, the entrytime (in human readable form), the entrytime (as
timestamp), a newline character, and a tab character.
The default format is: "%time -> %entry\\en". Don't forget the trailing 
newline.

.IP "-d <file> [<value>]"
Delete <value> from <file>. If <value> is omited, damnit will read <value>
from stdin (one per line), until either EOF or an empty line is detected.

.IP "-a <file> [<value>]"
Add <value> to <file>. If <value> is already stored in <file>, damnit will
only update the timestamp. If <value> is omited, damnit will read <value>
from stdin (one per line), until either EOF or an empty line is detected.

.IP "-l <file>"
List database file specified by <file>.

.SH DATABASE FILES
.B gnarwl
uses hashfiles for storing information on disk. Meaning, all
datasets consist of key and value pairs. The key is always 
a NULL terminated character string, while the value contains the timestamp,
when the key was entered (last time) into the file. This timestamp is of type
time_t (as returned by time(2)) and therefore, 
.B gnarwl's
database files are not copyable between different system architectures.

.SH AUTHOR
Patrick Ahlbrecht <p.ahlbrecht@billiton.de>

.SH SEE ALSO
.BR gnarwl(_MANSEC_)

.SH FILES
.I _HOMEDIR_/block/*
.RS
Every file in this directory represents an emailaddress, 
.B gnarwl
(already) received a mail for. Every time, gnarwl sends out an autorreply 
for an address, the recipient of that mail is locked into the according file.
.B gnarwl 
will not send any further autoreplies for this sender/receiver combo, until 
the timeout specified in gnarwl.cfg expires.
.RE
.P
.I _HOMEDIR_/blacklist.db
.RS
Emailaddresses listed as keys in this file are not subject to autoresponding 
(the addresses of root, postmaster, webmaster and the like should
be put herin). Note: The complete mailaddress, as it would appear in an email, 
must be specified here, as 
.B gnarwl
checks these "as-is".
.RE
.P
.I _HOMEDIR_/badheaders.db
.RS
Each entry in this file represent a line that may not occur in the header
of a received email. That is, 
.B gnarwl
won't reply to any mail, it is able
to match a headerline with an entry in this file.
.RE


doc/HISTORY
The versioning scheme of gnarwl consists of a major and a minor version
number. An increase in the former means a new feature was added, while an
increase in the later is to document a maintainance fix.

1.0:
  First public release.
  
1.1:
- Added a retrycounter and random sleep call to dbstuff.dbLock(). This should 
  fix an infinite loop problem of 1.0 where to instances of gnarwl would 
  try to access the same .db file over and over again, blocking each other.

- Updated documentation

- Added this file

- added README.pitfalls

- Renamed SECURITY README.security

- Updated Makefile (simplified it a bit)

- Removed hardcoded "sendmail" from  execl call in gnarwl.doReply(),
  replaced it with the result of basename(cfg.mta).

- openlog is now called once in main(), instead of immediatly before
  every syslog() call (what made me waste resources like this?).

- Removed exit(0) call in main(), return 0; schould do as well.  

- Updated gnarwl.printUsage(): Added version number, uses a single printf
  now.

- replaced exit(1) and exit(0) with exit(EXIT_FAILURE) and exit(EXIT_SUCCESS).
  Code should be more portable this way.

- gnarwl.addAddr() now checks first against cfg.maxmail before doing
  anything else.

- gnarwl.addAddr() now strips emailaddresses with angle brackets, e.g.:
  John Doe <john@doe.org> becomes john@doe.org
  
- Added new config switch "log_verbose", since -v cannot be specified in
  .forward files.

- closed some memory leaks


1.2
- Added a newline character to gnarwl.c (gcc 2.96 complained about it)
- Updates docs
  * fixed some typos
  * added a warning about openldap2.0.23 to INSTALL
  * added a manpage (bouyah!).
- Updated Makefile:
  * $(VER) is now computed from the directory basename, instead of being 
    hardcoded
  * added new variable $(CONFDIR)
  * The macros CFGFILE and VERSION are now defined via compiler switch
- Changed gnarwl.printUsage():
  * now also prints out default configfilename
  * versioninformation is taken from VERSION macro (instead of being
    hardcoded in the source).

2.0
- Renamed this file from CHANGES to HISTORY
- moved src/dbstuff.c to dbaccess.c
- killed dbaccess.dbExpired() ; integrated it's logic directly into dbCheck()
- added new management utility, damnit (yes, that's it's name).
- Splitted gnarwl.h -> every module has now it's own header
- Updates Makefile: Everything is now installed under /usr/local per default
- Changed gnarw.doReply() return type from int to void (since doReply always
  returned TRUE are bailed out itself if something went wrong).
- Reformulated most program messages
- More logging
- Base64 support. Actually it seems as if it was supported all the time, 
  but I didn't know of it #-) (So much for the documentation of openldap
  ;-). Why does ldapsearch return base64 encoded data, but 
  ldap_get_first_entry() not ?!).
- Changed return type of dbaccess.dbLock() from int to void. This method 
  will now exit(EXIT_FAILURE) directly if locking was unsucesfull.
- Fixed a bug in dbaccess.dbLock(): The do-while loop lacked the correct
  break condition, so it always looped 10 times.
- Replaced util.ewe(char*) with util.oom(). 
- Removed default away message (waste of space and didn't look pretty
  anyway).
- Added more filechecking to dbaccess.dbCheck(). It was rather weak
  before (would fail, if another process had the db file open for writing).
- Added flexible deny rules for mail.
- The blacklist is now stored in a gdbm file.
- Renamed bl_file to mail_blacklist
- Updated installation process
- Updated docs
- gnarwl.readLDAP() now uses ldap_init() instead of ldap_open()
- Added new version of billiton.schema
- Found a "nice feature": you may not free(2) the argument of basename(3).
  Fixed gnarwl.doReply accordingly. On the second thought: using basename
  was a stupid idea in the first place -> removed it completely.
- gnarwl is no longer bound to sendmail as MTA; MTA is now completly
  configureable (with parameters).
- renamed cfg.v_message to cfg.result
- added README.upgrade
- Updated manpage
- Disallowed autoreply for shared addresses.
- Renamed mail_limit to mail_limit_receivers, added mail_limit_header
- Merged README.pitfalls and README.security into FAQ
- Only the attributes specified in the config file will be queried
- Reworked gnarwl.readLDAP()
- added free(rep) after a call to gnarwl.doReply
- Removed v_begin and v_end, gnarwl now uses flexible macros
- Removed docs/USING
- gnarwl.readLDAP() now uses ldap_search_st instead of ldap_search_s
- Renamed config directive log_verbose to log_level.
- Eliminated "-v" commandlineswitch
- Added additional loglevels.
- Found a hefty bug in gnarwl.readHeader -> reimplemented it along with
  gnarwl.addAddr and ganrwl.main (they looked ugly anyway).
- added new function: dbaccess.dbOpen
- Moved declaration of "verbose" and "cfg" from gnarwl.c to conf.c, so
  damnit may link against the other objectfiles.
- Cleaned up the mess in the headerfiles:
  * Moved config structure from gnarwl.h to conf.h
  * All functions are now documentated in the corresponding headerfiles
  * gnarwl.h no longer includes all the other headers
- Added "contrib" branch for add-on software.

2.1
- Fixed a minor bug in dbaccess.dbCheck() and dbaccess.dbLock(): Both
  routines would ommit the trailing \0 of the key value. This was not
  critical, as both did so, but it was incompatible with damnit.
- Fixed a memory leak in damnit.listFile() (key.dptr and val.dptr are
  now freed).
- "make uninstall" will now also remove the system account of gnarwl (if
  present).
- Added a new macro "USERADD_ARG" to the Makefile.
- Updated docs
- Gnarwl.readHeader() would call util.cleanAddress without testing whether 
  the result was NULL or not (and therefore could segfault if one 
  supplied illegal from/to/cc or reply-to addresses) -> Added a test for
  NULL after every cleanAddress call.
- Changed places of main.doReply() and dbaccess.dbLock in gnarwl.main.

src/damnit.c
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <gdbm.h>
#include <errno.h>
#include <time.h>
#include <string.h>

#include <conf.h>
#include "dbaccess.h"
#include "config.h"
#include "util.h"

// How to display data
char *format="%time -> %entry\n";

/**
 * Print usage information
 */
void printUsage() {
  printf("\nDAMNIT(v%s) - DAtabase MAnagement InTerface\n\n\
  Damnit is a tool for listing/editing gnarwl's database files.\n\
  Options:\n\n\
  \t-h\t\t\t print usage information\n\
  \t-d <file> [<value>]\t delete <value> from <file>\n\
  \t-a <file> [<value>]\t add <value> to <file>\n\
  \t-l <file>\t\t list database file\n\
  \t-f <string>\t\t select output format (use with -l)\n\n",VERSION);
  exit(EXIT_SUCCESS);
}

/**
 * Print the contents of a db file (formated)
 * @param key key in the database
 * @param val value in the database
 */
void printFormated(const char* key, const time_t val) {
  char *tmp=NULL;
  char *tmp2=NULL;
  
  cpyStr(&tmp,format);
  cpyStr(&tmp2,ctime(&val));
  if (tmp2[strlen(tmp2)-1]=='\n') tmp2[strlen(tmp2)-1]='\0';
  
  expandVars(&tmp,"%entry",(char*)key);
  expandVars(&tmp,"%time",tmp2);
  expandVars(&tmp,"%tstamp","%d");
  expandVars(&tmp,"\\n","\n");
  expandVars(&tmp,"\\t","\t");
  
  /*
  free(tmp2);
  tmp2=(char*)calloc(strlen(tmp)+2,sizeof(char));
  if(tmp2==NULL) oom();
  strcpy(tmp2,tmp);
  if (tmp2[strlen(tmp2)]!='\n') tmp2[strlen(tmp2)]='\n';
  */
  printf(tmp,(int)val);
  free(tmp);
  free(tmp2);
}

/**
 * List the contents of a db file
 */
void listFile(char *name) {
  GDBM_FILE dbf;
  datum key;
  datum val;
  time_t t;
  
  dbf=dbOpen(name,GDBM_READER);
  
  if (dbf==NULL) {
    printf("IO Error: %s\n",name);
    exit(EXIT_FAILURE);
  }
    
  key = gdbm_firstkey ( dbf );
  while (key.dptr!=NULL) {
    val = gdbm_fetch(dbf,key);
    if (val.dsize==sizeof(time_t)) {
      memcpy(&t,val.dptr,sizeof(t));
      printFormated(key.dptr,t);
    }
    free(val.dptr);
    val.dptr=key.dptr;
    key = gdbm_nextkey(dbf,key);
    free(val.dptr);
  }
  dbClose(dbf);
}

void addToFile(char *fname, char *entry) {
  GDBM_FILE dbf;
  datum key;
  datum val;
  time_t t;
  
  dbf=dbOpen(fname,GDBM_WRCREAT);
  
  if (dbf==NULL) {
    printf("IO error: %s\n",fname);
    exit(EXIT_FAILURE);
  }

  t=time(NULL);
  key.dptr=entry;
  key.dsize=strlen(entry)+1;
  val.dptr=(char*)malloc(sizeof(t));
  if (val.dptr==NULL) {
    printf("Out of memory");
    exit(EXIT_FAILURE);
  }
  memcpy(val.dptr,&t,sizeof(t));
  val.dsize=sizeof(t);
  if (gdbm_store(dbf,key,val,GDBM_REPLACE)!=0) {
    printf("IO error: %s\n",fname);
    dbClose(dbf);
    exit(EXIT_FAILURE);
  }
  dbClose(dbf);
}

void delFromFile(char* fname, char* entry) {
  GDBM_FILE dbf;
  datum key;
  
  dbf=dbOpen(fname,GDBM_WRITER);
  
  if (dbf==NULL) {
    printf("IO error: %s\n",fname);
    exit(EXIT_FAILURE);
  }

  key.dptr=entry;
  key.dsize=strlen(entry)+1;
  if (!gdbm_exists(dbf,key)) {
    printf("Key not in database: %s\n",entry);
    exit(EXIT_FAILURE);
  }
  
  if (gdbm_delete(dbf,key)!=0) {
    printf("IO error: %s",fname);
    dbClose(dbf);
    exit(EXIT_FAILURE);
  }
  dbClose(dbf);
}

int main(int argc,char **argv) {
  int ch;
  char* fname=NULL;
  char* val=NULL;
  char buf[MAXLINE];
  
  setDefaults();
  
  if (argc==1) printUsage();
  
  while ((ch = getopt(argc, argv, "hf:l:a:d:")) != EOF) {
    switch((char)ch) {
      case 'h': printUsage();
      case 'f': {format=optarg;break;}
      case 'l': {
        listFile(optarg);
        exit(EXIT_SUCCESS);
      }
      case 'a': {
        fname=optarg;
        val=argv[optind];
        if (val!=NULL) addToFile(fname,val); 
        else while (fgets(buf,(int)sizeof(buf),stdin) && (*buf != '\n')) {
          if (buf[strlen(buf)-1]=='\n') buf[strlen(buf)-1]='\0';
          addToFile(fname,buf);
        }
        exit(EXIT_SUCCESS);
      }
      case 'd': {
        fname=optarg;
        val=argv[optind];
        if (val!=NULL) delFromFile(fname,val);
        else while (fgets(buf,(int)sizeof(buf),stdin) && (*buf != '\n')) {
          if (buf[strlen(buf)-1]=='\n') buf[strlen(buf)-1]='\0';
          delFromFile(fname,buf);
        }
        exit(EXIT_SUCCESS);
      }
    }
  }
  return EXIT_SUCCESS;
}
